File: ByteForge/src/compiler/lexer.cpp
#include "lexer.h"
#include <cctype>
#include <stdexcept>

namespace bf
{
    Lexer::Lexer(const std::string& source)
        : source(source), position(0), line(1), column(1)
    {
    }

    std::vector<Token> Lexer::tokenize()
    {
        std::vector<Token> tokens;

        while (!isAtEnd())
        {
            skipWhitespace();

            if (isAtEnd())
                break;

            char c = peek();

            if (std::isalpha(c) || c == '_')
            {
                tokens.push_back(identifier());
            }
            else if (std::isdigit(c))
            {
                tokens.push_back(number());
            }
            else
            {
                tokens.push_back(symbol());
            }
        }

        tokens.push_back(Token{TokenType::EndOfFile, "", line, column});
        return tokens;
    }

    bool Lexer::isAtEnd() const
    {
        return position >= source.size();
    }

    char Lexer::peek() const
    {
        return source[position];
    }

    char Lexer::advance()
    {
        char c = source[position++];
        if (c == '\n')
        {
            line++;
            column = 1;
        }
        else
        {
            column++;
        }
        return c;
    }

    void Lexer::skipWhitespace()
    {
        while (!isAtEnd())
        {
            char c = peek();
            if (c == ' ' || c == '\r' || c == '\t')
            {
                advance();
            }
            else if (c == '\n')
            {
                advance();
            }
            else
            {
                break;
            }
        }
    }

    Token Lexer::identifier()
    {
        int startCol = column;
        std::string value;

        while (!isAtEnd() && (std::isalnum(peek()) || peek() == '_'))
        {
            value += advance();
        }

        TokenType type = TokenType::Identifier;

        if (value == "let") type = TokenType::Let;
        else if (value == "print") type = TokenType::Print;
        else if (value == "while") type = TokenType::While;
        else if (value == "if") type = TokenType::If;
        else if (value == "else") type = TokenType::Else;

        return Token{type, value, line, startCol};
    }

    Token Lexer::number()
    {
        int startCol = column;
        std::string value;

        while (!isAtEnd() && std::isdigit(peek()))
        {
            value += advance();
        }

        return Token{TokenType::Number, value, line, startCol};
    }

    Token Lexer::symbol()
    {
        int startCol = column;
        char c = advance();

        switch (c)
        {
            case '+': return Token{TokenType::Plus, "+", line, startCol};
            case '-': return Token{TokenType::Minus, "-", line, startCol};
            case '*': return Token{TokenType::Star, "*", line, startCol};
            case '/': return Token{TokenType::Slash, "/", line, startCol};
            case '=':
                if (!isAtEnd() && peek() == '=')
                {
                    advance();
                    return Token{TokenType::EqualEqual, "==", line, startCol};
                }
                return Token{TokenType::Equal, "=", line, startCol};
            case '>': return Token{TokenType::Greater, ">", line, startCol};
            case '<': return Token{TokenType::Less, "<", line, startCol};
            case '(': return Token{TokenType::LeftParen, "(", line, startCol};
            case ')': return Token{TokenType::RightParen, ")", line, startCol};
            case '{': return Token{TokenType::LeftBrace, "{", line, startCol};
            case '}': return Token{TokenType::RightBrace, "}", line, startCol};
            case ';': return Token{TokenType::Semicolon, ";", line, startCol};
            default:
                throw std::runtime_error("Unexpected character: " + std::string(1, c));
        }
    }
}

File: Tokio-compiler/src/main/java/tokio/lexer/Lexer.java
package tokio.lexer;

import java.util.ArrayList;
import java.util.List;
import tokio.exceptions.LexerException;

public class Lexer {
    private final String source;
    private final List<Token> tokens;
    private int pos;
    private int line;
    private int column;

    public Lexer(String source) {
        this.source = source;
        this.tokens = new ArrayList<>();
        this.pos = 0;
        this.line = 1;
        this.column = 1;
    }

    public List<Token> tokenize() throws LexerException {
        while (!isAtEnd()) {
            char current = peek();
            switch (current) {
                case ' ':
                case '\r':
                case '\t':
                    advance();
                    break;
                case '\n':
                    advance();
                    line++;
                    column = 1;
                    break;
                case '+':
                    addToken(TokenType.PLUS);
                    advance();
                    break;
                case '-':
                    addToken(TokenType.MINUS);
                    advance();
                    break;
                case '*':
                    addToken(TokenType.STAR);
                    advance();
                    break;
                case '/':
                    addToken(TokenType.SLASH);
                    advance();
                    break;
                case '(':
                    addToken(TokenType.LPAREN);
                    advance();
                    break;
                case ')':
                    addToken(TokenType.RPAREN);
                    advance();
                    break;
                case '{':
                    addToken(TokenType.LBRACE);
                    advance();
                    break;
                case '}':
                    addToken(TokenType.RBRACE);
                    advance();
                    break;
                case ';':
                    addToken(TokenType.SEMICOLON);
                    advance();
                    break;
                case '=':
                    if (match('=')) {
                        addToken(TokenType.EQEQ);
                    } else {
                        addToken(TokenType.EQUAL);
                    }
                    break;
                case '!':
                    if (match('=')) {
                        addToken(TokenType.BANGEQ);
                    } else {
                        addToken(TokenType.BANG);
                    }
                    break;
                case '<':
                    addToken(match('=') ? TokenType.LTEQ : TokenType.LT);
                    break;
                case '>':
                    addToken(match('=') ? TokenType.GTEQ : TokenType.GT);
                    break;
                case '"':
                    string();
                    break;
                default:
                    if (isDigit(current)) {
                        number();
                    } else if (isAlpha(current)) {
                        identifier();
                    } else {
                        throw new LexerException("Unexpected character: " + current + " at line " + line + ", column " + column);
                    }
            }
        }
        addToken(TokenType.EOF);
        return tokens;
    }

    private void addToken(TokenType type) {
        tokens.add(new Token(type, null, line, column));
    }

    private void addToken(TokenType type, String lexeme) {
        tokens.add(new Token(type, lexeme, line, column));
    }

    private char advance() {
        pos++;
        column++;
        return source.charAt(pos - 1);
    }

    private boolean match(char expected) {
        if (isAtEnd()) return false;
        if (source.charAt(pos) != expected) return false;
        advance();
        return true;
    }

    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(pos);
    }

    private boolean isAtEnd() {
        return pos >= source.length();
    }

    private void string() throws LexerException {
        int start = pos + 1;
        while (!isAtEnd() && peek() != '"') {
            if (peek() == '\n') {
                line++;
                column = 1;
            }
            advance();
        }
        if (isAtEnd()) {
            throw new LexerException("Unterminated string at line " + line + ", column " + column);
        }
        advance();
        String value = source.substring(start, pos - 1);
        addToken(TokenType.STRING, value);
    }

    private void number() {
        int start = pos;
        while (!isAtEnd() && isDigit(peek())) advance();
        if (!isAtEnd() && peek() == '.') {
            advance();
            while (!isAtEnd() && isDigit(peek())) advance();
        }
        String value = source.substring(start, pos);
        addToken(TokenType.NUMBER, value);
    }

    private void identifier() {
        int start = pos;
        while (!isAtEnd() && isAlphaNumeric(peek())) advance();
        String value = source.substring(start, pos);
        TokenType type = TokenType.KEYWORDS.getOrDefault(value, TokenType.IDENTIFIER);
        addToken(type, value);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }
}
